### PR TITLE
feat(environment-info-screen): Environment Info Screen in Developer Menu

### PR DESCRIPTION
**Pull request by thisisabhash**
*Issue #, if available:* None

*Description of changes:*
1. Environment information screen - Developer environment info + Amplify plugins related changes
2. Versionable protocol to support showing amplify plugins version information
3. Changes in navigation bar title of device info screen

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by thisisabhash** [comment]
Screenshots
https://user-images.githubusercontent.com/23294582/87234344-8bba3080-c395-11ea-99f7-b6e03aea5263.png
https://user-images.githubusercontent.com/23294582/87234345-8bba3080-c395-11ea-970f-86536e414030.png
https://user-images.githubusercontent.com/23294582/87234348-8c52c700-c395-11ea-87b5-2e8535beedaf.png

**Comment by thisisabhash** [comment]
Updated screenshots

https://user-images.githubusercontent.com/23294582/87343561-abc32e80-c512-11ea-8566-a7fe8c98aebc.png
https://user-images.githubusercontent.com/23294582/87343562-ac5bc500-c512-11ea-9f25-88811d154334.png
https://user-images.githubusercontent.com/23294582/87343563-acf45b80-c512-11ea-838d-48002f7a997a.png
https://user-images.githubusercontent.com/23294582/87343566-ad8cf200-c512-11ea-85cb-f7cd7a9f0d24.png

File: Amplify/DevMenu/DevEnvironmentInfo.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+// struct to decode/encode information about developer environment in json format
+struct DevEnvironmentInfo: Codable {
+    let nodejsVersion, npmVersion, amplifyCLIVersion, podVersion: String
+    let xcodeVersion, osVersion: String
```
Comment (drochetti) [inline Amplify/DevMenu/DevEnvironmentInfo.swift:13]: prefer one declaration per line. We usually avoid declaring multiple fields per line.

File: Amplify/DevMenu/DevEnvironmentInfo.swift [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+    enum CodingKeys: String, CodingKey {
+        case nodejsVersion, npmVersion
+        case amplifyCLIVersion = "amplifyCliVersion"
+        case podVersion, xcodeVersion, osVersion
```
Comment (drochetti) [inline Amplify/DevMenu/DevEnvironmentInfo.swift:18]: formatting: one case per line

File: Amplify/DevMenu/View/DeviceInfoRow.swift [hunk 3]
```diff
@@ -7,6 +7,7 @@
 
 import SwiftUI
 
+/// View corresponding to each row in Device Information Screen
 @available(iOS 13.0.0, *)
 struct DeviceInfoRow: View {
```
Comment (drochetti) [inline Amplify/DevMenu/View/DeviceInfoRow.swift:12]: I wonder if `*InfoView` could be more generic to accommodate both Device and Environment info.

File: Amplify/DevMenu/View/EnvironmentInfoDetailView.swift [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    private let devEnvSectionTitle  = "Developer Environment Information"
+    private let notAvailable = "Not Available"
+
+    // Key descriptions for environment information
+    private let nodeJsVersionDesc = "Node.js version"
+    private let npmVersionDesc = "npm version"
+    private let amplifyCLIVersionDesc = "Amplify CLI version"
+    private let podVersionDesc = "CocoaPods version"
+    private let xcodeVersionDesc = "Xcode version"
+    private let osVersionDesc = "macOS version"
```
Comment (drochetti) [inline Amplify/DevMenu/View/EnvironmentInfoDetailView.swift:24]: Should these follow the same enum pattern you have for Device Info?

File: Amplify/DevMenu/View/EnvironmentInfoDetailView.swift [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            }
+        }
+        .navigationBarTitle(screenTitle)
+        .listStyle(GroupedListStyle())
+    }
+
+    private mutating func populateDevEnvSectionItems(devEnvInfo: DevEnvironmentInfo) {
+        devEnvSectionItems.append(EnvironmentInfoItem(
+            key: nodeJsVersionDesc,
+            value: devEnvInfo.nodejsVersion.isEmpty ? notAvailable : devEnvInfo.nodejsVersion))
```
Comment (drochetti) [inline Amplify/DevMenu/View/EnvironmentInfoDetailView.swift:64]: maybe `value` could be nullable and then the `notAvailable` logic happens inside the `InfoItem` view

File: Amplify/DevMenu/View/EnvironmentInfoRow.swift [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+struct EnvironmentInfoRow: View {
+    var rowItem: EnvironmentInfoItem
+
+    var body: some View {
+        VStack(alignment: .leading) {
+            Text(rowItem.key).bold()
+            Text(rowItem.value)
+        }
+    }
+}
```
Comment (drochetti) [inline Amplify/DevMenu/View/EnvironmentInfoRow.swift:21]: this view could be more generic, so both "device" and "environment" info could use it

File: Amplify/DevMenu/View/EnvironmentInfoDetailView.swift [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                        InfoRow(item: listItem)
+                    }
+                }
+            }
+        }
+        .navigationBarTitle(screenTitle)
+        .listStyle(GroupedListStyle())
+    }
+
+    private mutating func populateDevEnvSectionItems(devEnvInfo: DevEnvironmentInfo) {
```
Comment (royjit) [inline Amplify/DevMenu/View/EnvironmentInfoDetailView.swift:52]: Lets move these data related function out from UI. Create a class similar to my example above. `EnvironmentInfoHelper?`

File: Amplify/DevMenu/View/EnvironmentInfoDetailView.swift [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+            value: devEnvInfo.podVersion))
+        devEnvSectionItems.append(EnvironmentInfoItem(
+            key: EnvironmentInfoItemType.xcodeVersion.key,
+            value: devEnvInfo.xcodeVersion))
+        devEnvSectionItems.append(EnvironmentInfoItem(
+            key: EnvironmentInfoItemType.osVersion.key,
+            value: devEnvInfo.osVersion))
+    }
+
+    private mutating func populateAmplifyEnvSectionItems() {
```
Comment (royjit) [inline Amplify/DevMenu/View/EnvironmentInfoDetailView.swift:73]: Move this to a separate struct, something like: `AmplifyEnvironmentInfoHelper`

File: Amplify/DevMenu/View/InfoRow.swift [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    init(item: EnvironmentInfoItem) {
+        self.key = item.key
+        if item.value.isEmpty {
+            self.value = notAvailable
+        } else {
+            self.value = item.value
+        }
+    }
```
Comment (royjit) [inline Amplify/DevMenu/View/InfoRow.swift:33]: As mentioned in the previous comment please check if you can make this generic using a protocol.

**Comment by thisisabhash** [comment]
https://user-images.githubusercontent.com/23294582/87457428-f6ee4780-c5cd-11ea-8355-e8168cdf63c1.png

File: Amplify/DevMenu/Data/DeviceInfoItemType.swift [hunk 10]
```diff
@@ -6,52 +6,13 @@
...
-            return UIDevice.current.localizedModel
-        case .isSimulator:
-            #if targetEnvironment(simulator)
-            return "Yes"
-            #else
-            return "No"
-            #endif
-        }
-    }
+    case deviceName(String?)
```
Comment (royjit) [inline Amplify/DevMenu/Data/DeviceInfoItemType.swift:12]: Can you please add comments to these case. What is the difference between deviceName and systemName? Maybe add comments with an example associated value. For ex:

```
enum DeviceInfoItemType {
  /// Device name represents the name of the device it is running.
  /// 
  /// For example if the app is running in a device named "Jithin's iPhone" the associated value of
  /// .deviceName will be "Jithin's iPhone".
  case deviceName(String?)
}
```

File: Amplify/DevMenu/Data/EnvironmentInfoItem.swift [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+/// Data class for  each item showing Developer Environment Information
+@available(iOS 13.0.0, *)
+struct EnvironmentInfoItem: Identifiable, InfoItemProvider {
+
+    let id = UUID()
+    let type: EnvironmentInfoItemType
+    let notAvailable: String = "Not available"
+
+    init(type: EnvironmentInfoItemType) {
```
Comment (royjit) [inline Amplify/DevMenu/Data/EnvironmentInfoItem.swift:18]: init is required? I think the struct create init methods for us.

File: Amplify/DevMenu/Data/PluginInfoItem.swift [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+@available(iOS 13.0.0, *)
+struct PluginInfoItem: Identifiable, InfoItemProvider {
+
+    let id = UUID()
+    var displayName: String
+    var information: String
+
+    init(displayName: String, information: String) {
+        self.displayName = displayName.isEmpty ? "Unknown Plugin" : displayName
+        self.information = information.isEmpty ? "Not Available" : information
```
Comment (royjit) [inline Amplify/DevMenu/Data/PluginInfoItem.swift:19]: Let us move the string constants to a separate file "DevMenuStringConstants". I see some strings reused like "Not Available". Example - https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/Auth/AWSCognitoAuthPlugin/Support/Constants/AuthPluginErrorConstants.swift

**Pull request merged**